### PR TITLE
JENKINS-36376# Adding missing capabilities

### PR DESCRIPTION
**Pull request by vivek**
# Description

See https://issues.jenkins-ci.org/browse/JENKINS-36376.

# Submitter checklist
- [x] Link to JIRA ticket in description, if appropriate.
- [x] Change is code complete and matches issue description
- [x] Apppropriate unit or acceptance tests or explaination to why this change has no tests
- [x] Reviewer's manual test instructions provided in PR description. See Reviewer's first task below.

Call classes API: 

GET http://localhost:8080/jenkins/blue/rest/classes/?q=io.jenkins.blueocean.service.embedded.rest.AbstractPipelineImpl,io.jenkins.blueocean.rest.impl.pipeline.PipelineImpl,io.jenkins.blueocean.rest.impl.pipeline.BranchImpl

- For class io.jenkins.blueocean.service.embedded.rest.AbstractPipelineImpl, it's capabilities should include hudson.model.Job but not io.jenkins.blueocean.rest.model.BlueBranch and org.jenkinsci.plugins.workflow.job.WorkflowJob

- For class io.jenkins.blueocean.rest.impl.pipeline.PipelineImpl, it's capabilities should include hudson.model.Job, org.jenkinsci.plugins.workflow.job.WorkflowJob but not io.jenkins.blueocean.rest.model.BlueBranch.

- For class io.jenkins.blueocean.rest.impl.pipeline.BranchImpl, it's capabilities should include hudson.model.Job, io.jenkins.blueocean.rest.model.BlueBranch and org.jenkinsci.plugins.workflow.job.WorkflowJob

@cliffmeyers  Please review. Note: there is capability to indicate it's a pipeline job (org.jenkinsci.plugins.workflow.job.WorkflowJob), capability to indicate a pipeline is branch (io.jenkins.blueocean.rest.model.BlueBranch). pipeline with neither is some other jenkins job type.

- [x] Ran Acceptance Test Harness against PR changes.

https://ci.blueocean.io/job/Acceptance%20Tests%20Param/15/

# Reviewer checklist
- [x] Run the changes and verified the change matches the issue description
- [x] Reviewed the code
- [x] Verified that the appropriate tests have been written or valid explaination given

@jenkinsci/code-reviewers @reviewbybees 

- Refactored pipeline factory to resolve WorkflowJob and moved to pipeline module
- Capability added to branch, pipeline, non-pipeline jobs etc.

### CODE REVIEW

File: blueocean-rest-impl/src/main/java/io/jenkins/blueocean/service/embedded/rest/PipelineFolderImpl.java [hunk 1]
```diff
@@ -97,7 +97,7 @@ public BlueFavorite favorite(@JsonBody BlueFavoriteAction favoriteAction) {
 
     @Override
     public Link getLink() {
-        return OrganizationImpl.INSTANCE.getLink().rel("pipelines").rel(PipelineImpl.getRecursivePathFromFullName(this));
+        return OrganizationImpl.INSTANCE.getLink().rel("pipelines").rel(AbstractPipelineImpl.getRecursivePathFromFullName(this));
```
Comment (michaelneale) [inline blueocean-rest-impl/src/main/java/io/jenkins/blueocean/service/embedded/rest/PipelineFolderImpl.java]: I am sure I have asked before, but how can something be both abstract and a concrete impl at the same time?

**Comment by michaelneale** [comment]
Code looks nice, and tests, I'll let @cliffmeyers eyeball this to ensure it seems right functionally 

"AbstractPipelineImpl" still reads funny to me as a name ;) hehehe

File: blueocean-rest-impl/src/main/java/io/jenkins/blueocean/service/embedded/rest/PipelineFolderImpl.java (hunk 1 above)
Comment (vivek) [inline blueocean-rest-impl/src/main/java/io/jenkins/blueocean/service/embedded/rest/PipelineFolderImpl.java]: @michaelneale Abstract in the class name is different from Java 'abstract' keyword. Abstract in class name means its not specific to particular implementation and serves most generic jobs.

Comment (michaelneale) [inline blueocean-rest-impl/src/main/java/io/jenkins/blueocean/service/embedded/rest/PipelineFolderImpl.java]: Gotcha. "GenericPipelineImpl" may have also worked? (irrelevant to this PR, but just curious). I find naming things really hard.

Comment (vivek) [inline blueocean-rest-impl/src/main/java/io/jenkins/blueocean/service/embedded/rest/PipelineFolderImpl.java]: Maybe, I find it hard too, we have precedence on doing it with AbstractRunImpl as well. I have seen such usage elsewhere, such names are really opinion oriented, no hard and fast. For the same reason, you could have an 'abstract' class without Abstract name in it:)

**Comment by cliffmeyers** [comment]
@vivek so should I expect that a Freestyle job would be returned as "AbstractPipelineImpl" which when fetched via Capabilities API would yield only the "Job" capability?

**Pull request merged**